### PR TITLE
Added a Pebble function [uniq] to deduplice array #6417

### LINKED ISSUE #6417: Add a Pebble function [uniq] to deduplice array
### Feature description

Add a Pebble function [uniq] to deduplice array

[1,2,2,3] | uniq -> [1,2,3]

OR

Adding a bool parameter to merge with uniq

[1,2] | merge([2,3],true) -> [1,2,3]

### PR DESCRIPTION
**Pull request by Rajarajan985**
Add a Pebble function [uniq] to deduplice array in the code to convert duplicate value to unique values
[1,2,2,3] | uniq -> [1,2,3]

Added new file:
core/src/main/java/io/kestra/core/runners/pebble/filters/UniqFilter.java

Changes added in existing file:
core/src/main/java/io/kestra/core/docs/DocumentationGenerator.java

Fixes https://github.com/kestra-io/kestra/issues/6417

### CODE REVIEW

**Comment by Rajarajan985** [comment]
Hi @loicmathieu  @MilosPaunovic ,
I made the changes to remove the duplicate values in the list
can check, is there any changes are required for this code

**Comment by MilosPaunovic** [comment]
@Rajarajan985 Can you add a test for this?

**Comment by Rajarajan985** [comment]
> @Rajarajan985 Can you add a test for this?

@MilosPaunovic I'm working on it.

**Comment by Rajarajan985** [comment]
Hi @MilosPaunovic,
Added the test case for uniqFIlter, Can you verify once. 
It is taking mush time to execute this command in my system "gradle test --tests "UniqFilterTest" (Can you describe how to verify test case are working fine)
May I know below mentioned files i haved to add anything related to uniqFilter function
index.peb, macro.peb, task.peb these are available in the pebbleEnigne adding funtionality.

**Comment by MilosPaunovic** [comment]
@loicmathieu is the one which will review this PR, when he gets the chance.

File: core/src/main/java/io/kestra/core/runners/pebble/filters/UniqFilter.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        // Check if the input is a list
+        if (input instanceof List<?>) {
+            List<?> list = (List<?>) input;
+
+            // Deduplicate the list by using distinct stream operation
+            return list.stream().distinct().collect(Collectors.toList());
+        }
+
+        // If the input is not a list, just return it as it is
+        return input;
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/pebble/filters/UniqFilter.java]: You should throw a `PebbleException` here as we prefer to be strict

File: core/src/main/java/io/kestra/core/runners/pebble/filters/UniqFilter.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+public class UniqFilter implements Filter {
+
+    @Override
+    public List<String> getArgumentNames() {
+        // No arguments are needed for the uniq filter
+        return null;
+    }
+
+    @Override
+    public Object apply(Object input, Map<String, Object> args, PebbleTemplate self, EvaluationContext context,
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/pebble/filters/UniqFilter.java]: If the input is null, please return null to avoid possible NullPointerException.

File: core/src/test/java/io/kestra/core/runners/pebble/filters/UniqFilterTest.java
Comment (loicmathieu) [inline core/src/test/java/io/kestra/core/runners/pebble/filters/UniqFilterTest.java:1]: We use our variable renderer to test functions closer to how we use them.
You can take as inspiration for ex io.kestra.core.runners.pebble.filters.ToJsonFilterTest

**Review by loicmathieu — CHANGES_REQUESTED** [review]

**Comment by loicmathieu** [comment]
Please also rename the filter to `distinct`.

File: core/src/test/java/io/kestra/core/runners/pebble/filters/UniqFilterTest.java
Comment (Rajarajan985) [reply core/src/test/java/io/kestra/core/runners/pebble/filters/UniqFilterTest.java:1]: @loicmathieu Yes sure, I'm working on it.

**Comment by Rajarajan985** [comment]
@loicmathieu I have made the changes.
While running the test cases getting this below exception Filter [distinct] does not exist.. 
Can you help me to sort out this issue.

2024-12-22 16:15:20,433 ERROR Test worker  io.kestra.core.secret.SecretService Could not decode secret 'SECRET_NON_B64_SECRET', make sure it is Base64-encoded: Illegal base64 character 20
Before applying distinct filter : ["one","two","one","three","four","five","three",1,2,3,1,2,2,1.123,2.123,1.123,10.0,10.0]

io.pebbletemplates.pebble.error.PebbleException: Filter [distinct] does not exist. ({{ vars.second.list | distinct }}:1)
io.kestra.core.exceptions.IllegalVariableEvaluationException: io.pebbletemplates.pebble.error.PebbleException: Filter [distinct] does not exist. ({{ vars.second.list | distinct }}:1)
	at io.kestra.core.runners.VariableRenderer.properPebbleException(VariableRenderer.java:64)
	at io.kestra.core.runners.VariableRenderer.renderOnce(VariableRenderer.java:120)
	at io.kestra.core.runners.VariableRenderer.render(VariableRenderer.java:91)
	at io.kestra.core.runners.VariableRenderer.render(VariableRenderer.java:76)
	at io.kestra.core.runners.VariableRenderer.render(VariableRenderer.java:68)
	at io.kestra.core.runners.pebble.filters.DistinctFilterTest.toDistinctFilter(DistinctFilterTest.java:62)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension$2.proceed(MicronautJunit5Extension.java:142)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:162)
	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptTest(AbstractMicronautExtension.java:119)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.interceptTestMethod(MicronautJunit5Extension.java:129)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: io.pebbletemplates.pebble.error.PebbleException: Filter [distinct] does not exist. ({{ vars.second.list | distinct }}:1)
	at io.pebbletemplates.pebble.node.expression.FilterExpression.evaluate(FilterExpression.java:49)
	at io.pebbletemplates.pebble.node.PrintNode.render(PrintNode.java:37)
	at io.pebbletemplates.pebble.node.BodyNode.render(BodyNode.java:44)
	at io.pebbletemplates.pebble.node.RootNode.render(RootNode.java:31)
	at io.pebbletemplates.pebble.template.PebbleTemplateImpl.evaluate(PebbleTemplateImpl.java:157)
	at io.pebbletemplates.pebble.template.PebbleTemplateImpl.evaluate(PebbleTemplateImpl.java:96)
	at io.kestra.core.runners.VariableRenderer.renderOnce(VariableRenderer.java:114)
	... 11 more


Filter [distinct] does not exist. ({{ vars.second.list | distinct }}:1)
io.pebbletemplates.pebble.error.PebbleException: Filter [distinct] does not exist. ({{ vars.second.list | distinct }}:1)
	at app//io.pebbletemplates.pebble.node.expression.FilterExpression.evaluate(FilterExpression.java:49)
	at app//io.pebbletemplates.pebble.node.PrintNode.render(PrintNode.java:37)
	at app//io.pebbletemplates.pebble.node.BodyNode.render(BodyNode.java:44)
	at app//io.pebbletemplates.pebble.node.RootNode.render(RootNode.java:31)
	at app//io.pebbletemplates.pebble.template.PebbleTemplateImpl.evaluate(PebbleTemplateImpl.java:157)
	at app//io.pebbletemplates.pebble.template.PebbleTemplateImpl.evaluate(PebbleTemplateImpl.java:96)
	at app//io.kestra.core.runners.VariableRenderer.renderOnce(VariableRenderer.java:114)
	at app//io.kestra.core.runners.VariableRenderer.render(VariableRenderer.java:91)
	at app//io.kestra.core.runners.VariableRenderer.render(VariableRenderer.java:76)
	at app//io.kestra.core.runners.VariableRenderer.render(VariableRenderer.java:68)
	at app//io.kestra.core.runners.pebble.filters.DistinctFilterTest.toDistinctFilter(DistinctFilterTest.java:62)
	at java.base@21.0.5/java.lang.reflect.Method.invoke(Method.java:580)
	at app//io.micronaut.test.extensions.junit5.MicronautJunit5Extension$2.proceed(MicronautJunit5Extension.java:142)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:162)
	at app//io.micronaut.test.extensions.AbstractMicronautExtension.interceptTest(AbstractMicronautExtension.java:119)
	at app//io.micronaut.test.extensions.junit5.MicronautJunit5Extension.interceptTestMethod(MicronautJunit5Extension.java:129)
	at java.base@21.0.5/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base@21.0.5/java.util.ArrayList.forEach(ArrayList.java:1596)

**Comment by Rajarajan985** [comment]
@loicmathieu I have made the all the changes.
Now test cases are passed.
Can you verify at your end once.

Please find the below test case passed screenshot for your reference.
Capture

File: core/src/main/java/io/kestra/core/runners/pebble/filters/DistinctFilter.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+public class DistinctFilter implements Filter {
+
+    @Override
+    public List<String> getArgumentNames() {
+        return null;
+    }
+
+    @Override
+    public Object apply(Object input, Map<String, Object> args, PebbleTemplate self, EvaluationContext context,
+                        int lineNumber) throws PebbleException, NullPointerException {
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/pebble/filters/DistinctFilter.java]:
```suggestion
                        int lineNumber) throws PebbleException {
```

Please don't declare throws NullPointerException, NPE are always considered as a bug.

File: core/src/main/java/io/kestra/core/runners/pebble/filters/DistinctFilter.java
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/runners/pebble/filters/DistinctFilter.java:1]: Can you allow null on this filter?

Please add this in the beginning of the method:

```java
        if (input == null) {
            return null;
        }
```

File: core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                    "string", "string",
+                    "int", 1,
+                    "float", 1.123F
+                )
+            ))
+        );
+
+        //Test rendering the list without the distinct filter
+        String render = variableRenderer.render("{{ vars.second.list }}", vars);
+        System.out.println("Before applying distinct filter : " + render);
```
Comment (loicmathieu) [inline core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java]:
```suggestion
```

File: core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        //Verify that the list contains duplicates
+        assertThat(render, containsString("one"));
+        assertThat(render, containsString("two"));
+        assertThat(render, containsString("three"));
+        assertThat(render, containsString("1"));
+        assertThat(render, containsString("1.123"));
+
+        //Apply the distinct filter
+        String distinctRender = variableRenderer.render("{{ vars.second.list | distinct }}", vars);
+        System.out.println("After applying distinct filter : " + distinctRender);
```
Comment (loicmathieu) [inline core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java]:
```suggestion
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]

**Comment by Rajarajan985** [comment]
@loicmathieu I have made the changes. Can you verify once

File: core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        assertThat(distinctRender, startsWith("["));
+        assertThat(distinctRender, endsWith("]"));
+
+        //Edge case: an empty list
+        render = variableRenderer.render("{{ [] | distinct }}", Map.of());
+        assertThat(render, is("[]"));
+    }
+
+    @Test
+    void distinctFilterWithNullInput() {
```
Comment (loicmathieu) [inline core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java]: I think this test will now pass.

**Review by loicmathieu — APPROVED** [review]
LGTM, thanks a lot.

**Comment by Rajarajan985** [comment]
@loicmathieu Some check were unsuccessful
Can you check, is there any changes I have to make

File: core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java (hunk 6 above)
Comment (loicmathieu) [reply core/src/test/java/io/kestra/core/runners/pebble/filters/DistinctFilterTest.java]: I was not clear, this test would not pass as null is allowed.
Can you update the test so we can merge?

**Comment by loicmathieu** [comment]
@Rajarajan96597 yes, the test `distinctFilterWithNullInput` is now wrong as it assert that an exception is thrown but as we return null for null parameter it no longer throws an exception, please update the test accordingly

**Comment by Rajarajan985** [comment]
@loicmathieu Done the changes for the test cases.
Can you verify the test case once.

**Comment by Rajarajan985** [comment]
@loicmathieu At my end have to raise any merge request for this code like "Update with merge commit" or "Update with rebase" to commit code in the branch

**Review by loicmathieu — APPROVED** [review]
